File: 程序员面试金典/4.5合法二叉搜索树/isValidBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// 中序遍历法
class Solution {
    private Integer last = null;

    public boolean isValidBST(TreeNode root) {
        if (root == null)
            return true;
        if (!isValidBST(root.left))
            return false;
        if (last != null && root.val <= last)
            return false;
        last = root.val;
        if (!isValidBST(root.right))
            return false;
        return true;
    }
}

File: Algorithms/98/isValidBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# 中序遍历
class Solution:
    def __init__(self):
        self.prev = float('-inf')
        self.ans = True

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        self.inorder(root)
        return self.ans

    def inorder(self, root):
        if not root:
            return
        self.inorder(root.left)
        if self.prev >= root.val:
            self.ans = False
        self.prev = root.val
        self.inorder(root.right)
